File: java/src/main/java/global/thalion/ttio/genomics/ReferenceImport.java
/*
 * TTI-O Java Implementation
 * Copyright (c) 2026 The Thalion Initiative
 * SPDX-License-Identifier: Apache-2.0
 */
package global.thalion.ttio.genomics;

import global.thalion.ttio.Enums;
import global.thalion.ttio.SpectralDataset;
import global.thalion.ttio.providers.StorageDataset;
import global.thalion.ttio.providers.StorageGroup;
import global.thalion.ttio.providers.StorageProvider;

import java.io.ByteArrayOutputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Reference-FASTA value class staged for embedding into a {@code .tio}
 * container.
 *
 * <p>A {@code ReferenceImport} is the parsed result of a reference-FASTA
 * file (many short or long chromosome records, no quality scores). It
 * carries the chromosome names, per-chromosome sequence bytes, and a
 * content-MD5 suitable for the {@code @md5} attribute on
 * {@code /study/references/<uri>/} groups.</p>
 *
 * <p><b>Cross-language equivalents:</b> Python
 * {@code ttio.genomic.reference_import.ReferenceImport}, Objective-C
 * {@code TTIOReferenceImport}.</p>
 *
 * <p>The MD5 algorithm sorts chromosomes by name (so the digest is
 * order-invariant), then concatenates the per-chromosome
 * {@code sequence_bytes} verbatim (case-preserving, no framing) and
 * digests the result. Cross-language byte-equal. Unified in v1.1.0
 * with the REF_DIFF_V2 auto-embed writer's stamp; the previous
 * name-framed form is gone.</p>
 */
public final class ReferenceImport {

    private final String uri;
    private final List<String> chromosomes;
    private final List<byte[]> sequences;
    private final byte[] md5;

    /**
     * Construct a reference import. Computes MD5 from the chromosome
     * set if {@code md5} is {@code null}.
     *
     * @param uri          reference URI (e.g. {@code "GRCh38.p14"}).
     * @param chromosomes  chromosome names in FASTA file order.
     * @param sequences    per-chromosome sequence bytes (case-preserving).
     * @param md5          16-byte content MD5, or {@code null} to compute.
     */
    public ReferenceImport(
        String uri, List<String> chromosomes, List<byte[]> sequences, byte[] md5
    ) {
        this.uri = Objects.requireNonNull(uri, "uri");
        Objects.requireNonNull(chromosomes, "chromosomes");
        Objects.requireNonNull(sequences, "sequences");
        if (chromosomes.size() != sequences.size()) {
            throw new IllegalArgumentException(
                "chromosomes / sequences length mismatch: "
                    + chromosomes.size() + " vs " + sequences.size()
            );
        }
        this.chromosomes = List.copyOf(chromosomes);
        // sequences kept as a read-only view of the originals; copy
        // the list but not the byte arrays.
        this.sequences = Collections.unmodifiableList(new ArrayList<>(sequences));
        if (md5 == null) {
            this.md5 = computeMd5(this.chromosomes, this.sequences);
        } else {
            if (md5.length != 16) {
                throw new IllegalArgumentException(
                    "md5 must be 16 bytes, got " + md5.length
                );
            }
            this.md5 = md5.clone();
        }
    }

    /** Convenience constructor that always computes the MD5 from the
     *  chromosome set. */
    public ReferenceImport(
        String uri, List<String> chromosomes, List<byte[]> sequences
    ) {
        this(uri, chromosomes, sequences, null);
    }

    /**
     * Compute the canonical content-MD5 over a chromosome set. The
     * algorithm sorts by name (order-invariant), then concatenates
     * the per-chromosome sequence bytes verbatim (case-preserving,
     * no framing) into an MD5 digest. Matches the REF_DIFF_V2
     * auto-embed writer's stamp byte-for-byte (unified in v1.1.0).
     */
    public static byte[] computeMd5(List<String> chromosomes, List<byte[]> sequences) {
        if (chromosomes.size() != sequences.size()) {
            throw new IllegalArgumentException(
                "chromosomes / sequences length mismatch: "
                    + chromosomes.size() + " vs " + sequences.size()
            );
        }
        // Build a (name, seq) list and sort by name.
        Map<String, byte[]> indexByName = new LinkedHashMap<>();
        for (int i = 0; i < chromosomes.size(); i++) {
            indexByName.put(chromosomes.get(i), sequences.get(i));
        }
        List<String> sortedNames = new ArrayList<>(indexByName.keySet());
        Collections.sort(sortedNames);
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            for (String name : sortedNames) {
                md.update(indexByName.get(name));
            }
            return md.digest();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("MD5 unavailable on this JVM", e);
        }
    }

    /** Reference URI. */
    public String uri() { return uri; }

    /** Chromosome names in FASTA file order. */
    public List<String> chromosomes() { return chromosomes; }

    /**
     * Per-chromosome sequence bytes (case-preserving). Returned in
     * FASTA file order; positionally aligned with
     * {@link #chromosomes()}.
     */
    public List<byte[]> sequences() { return sequences; }

    /** 16-byte content MD5. Returns a defensive copy. */
    public byte[] md5() { return md5.clone(); }

    /** Sum of sequence lengths across all chromosomes. */
    public long totalBases() {
        long n = 0;
        for (byte[] s : sequences) n += s.length;
        return n;
    }

    /**
     * Look up a chromosome's sequence by name.
     *
     * @throws java.util.NoSuchElementException if not present.
     */
    public byte[] chromosome(String name) {
        for (int i = 0; i < chromosomes.size(); i++) {
            if (chromosomes.get(i).equals(name)) {
                return sequences.get(i);
            }
        }
        List<String> known = new ArrayList<>(chromosomes);
        Collections.sort(known);
        throw new java.util.NoSuchElementException(
            "chromosome '" + name + "' not present in reference '"
                + uri + "' (known: " + known + ")"
        );
    }

    /** Returns the lowercase-hex form of the content MD5. */
    public String md5Hex() {
        StringBuilder sb = new StringBuilder(32);
        for (byte b : md5) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    /**
     * Read an embedded reference from {@code /study/references/<uri>/}.
     *
     * <p>Layout (matches the writer in
     * {@code SpectralDataset.embedReferencesForRuns}):</p>
     * <ul>
     *   <li>{@code refGroup} attribute {@code reference_uri} = the
     *       reference URI; falls back to {@code refGroup.name()}.</li>
     *   <li>{@code refGroup} attribute {@code md5} = lowercase-hex
     *       content MD5; preserved verbatim into the returned
     *       {@code ReferenceImport} so byte-for-byte round-trip is
     *       maintained.</li>
     *   <li>{@code refGroup/chromosomes/} = sub-group containing one
     *       child per chromosome.</li>
     *   <li>{@code refGroup/chromosomes/<name>/data} = UINT8 dataset of
     *       sequence bytes (case-preserving).</li>
     * </ul>
     *
     * <p>Chromosomes are returned in the order
     * {@link StorageGroup#childNames()} reports them — the writer
     * sorts alphabetically before persisting, so for any file written
     * by this library the order is alphabetic.</p>
     *
     * @param refGroup the {@code /study/references/<uri>/} group
     * @return a fully-populated {@code ReferenceImport}
     * @since 1.1.0
     */
    public static ReferenceImport readFromGroup(StorageGroup refGroup) {
        Objects.requireNonNull(refGroup, "refGroup");

        // URI: prefer @reference_uri, fall back to the group name.
        String uri;
        if (refGroup.hasAttribute("reference_uri")) {
            Object v = refGroup.getAttribute("reference_uri");
            uri = v != null ? v.toString() : refGroup.name();
        } else {
            uri = refGroup.name();
        }

        // MD5: preserve verbatim from @md5 (lowercase hex) when
        // present, so the read-back ReferenceImport carries the same
        // digest bytes as the writer used.
        byte[] md5 = null;
        if (refGroup.hasAttribute("md5")) {
            Object v = refGroup.getAttribute("md5");
            if (v != null) {
                md5 = parseHexLocal(v.toString());
            }
        }

        List<String> chromNames = new ArrayList<>();
        List<byte[]> seqs = new ArrayList<>();
        try (StorageGroup chromsGrp = refGroup.openGroup("chromosomes")) {
            for (String name : chromsGrp.childNames()) {
                try (StorageGroup chromGrp = chromsGrp.openGroup(name)) {
                    try (StorageDataset ds = chromGrp.openDataset("data")) {
                        byte[] bytes = (byte[]) ds.readAll();
                        chromNames.add(name);
                        seqs.add(bytes);
                    }
                }
            }
        }

        return new ReferenceImport(uri, chromNames, seqs, md5);
    }

    /**
     * Embed this reference at {@code /study/references/<uri>/} inside
     * {@code dataset}'s open storage backing.
     *
     * <p>Layout (cross-language byte-equal — matches Python's
     * {@code ReferenceImport.write_to_dataset} and the canonical
     * embed-helper writer used by {@code embedReference=true}
     * runs):</p>
     *
     * <ul>
     *   <li>{@code /study/references/<uri>/} group with
     *       {@code @md5} (32-char lowercase hex) and
     *       {@code @reference_uri} attributes.</li>
     *   <li>{@code chromosomes/<name>/} sub-group per chromosome,
     *       in alphabetic order, with an {@code @length} (int64)
     *       attribute.</li>
     *   <li>{@code chromosomes/<name>/data} UINT8 ZLIB-compressed
     *       dataset of sequence bytes.</li>
     * </ul>
     *
     * <p>If a reference with the same {@code uri} is already
     * embedded and {@code overwrite} is {@code false}, throws
     * {@link IllegalStateException} (mirrors Python's
     * {@code FileExistsError}). When {@code overwrite} is
     * {@code true}, the existing group is deleted first.</p>
     *
     * @param dataset   open dataset; must expose a writable
     *                  {@link StorageProvider} via
     *                  {@link SpectralDataset#provider()}.
     * @param overwrite if {@code true}, replace any existing
     *                  reference under the same URI.
     * @throws IllegalStateException if {@code overwrite=false} and a
     *         reference with the same URI is already embedded, or if
     *         the dataset has no open writable provider.
     * @since 1.1.0
     */
    public void writeToDataset(SpectralDataset dataset, boolean overwrite) {
        Objects.requireNonNull(dataset, "dataset");
        StorageProvider provider = dataset.provider();
        if (provider == null || !provider.isOpen()) {
            throw new IllegalStateException(
                "ReferenceImport.writeToDataset requires an open dataset "
                + "with a writable provider; got "
                + (provider == null ? "null provider" : "closed provider")
                + ".");
        }
        try (StorageGroup root = provider.rootGroup()) {
            StorageGroup study;
            if (root.hasChild("study")) {
                study = root.openGroup("study");
            } else {
                study = root.createGroup("study");
            }
            try (StorageGroup ignoredStudy = study) {
                StorageGroup refsGrp;
                if (study.hasChild("references")) {
                    refsGrp = study.openGroup("references");
                } else {
                    refsGrp = study.createGroup("references");
                }
                try (StorageGroup ignoredRefs = refsGrp) {
                    if (refsGrp.hasChild(uri)) {
                        if (!overwrite) {
                            throw new IllegalStateException(
                                "reference '" + uri + "' already embedded "
                                + "at /study/references/" + uri + "; "
                                + "pass overwrite=true to replace.");
                        }
                        refsGrp.deleteChild(uri);
                    }
                    try (StorageGroup refGrp = refsGrp.createGroup(uri)) {
                        refGrp.setAttribute("md5", md5Hex());
                        refGrp.setAttribute("reference_uri", uri);
                        try (StorageGroup chromsGrp = refGrp.createGroup("chromosomes")) {
                            // Build a (name -> seq) map and sort by
                            // name so the on-disk child order matches
                            // the canonical embed-helper writer
                            // byte-for-byte.
                            Map<String, byte[]> byName = new LinkedHashMap<>();
                            for (int i = 0; i < chromosomes.size(); i++) {
                                byName.put(chromosomes.get(i), sequences.get(i));
                            }
                            List<String> sortedNames = new ArrayList<>(byName.keySet());
                            Collections.sort(sortedNames);
                            for (String chromName : sortedNames) {
                                byte[] seq = byName.get(chromName);
                                try (StorageGroup c = chromsGrp.createGroup(chromName)) {
                                    c.setAttribute("length", (long) seq.length);
                                    StorageDataset ds;
                                    try {
                                        ds = c.createDataset("data",
                                            Enums.Precision.UINT8, seq.length,
                                            65536, Enums.Compression.ZLIB, 6);
                                    } catch (UnsupportedOperationException e) {
                                        ds = c.createDataset("data",
                                            Enums.Precision.UINT8, seq.length,
                                            0, Enums.Compression.NONE, 0);
                                    }
                                    try (StorageDataset closeMe = ds) {
                                        closeMe.writeAll(seq);
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    /**
     * Convenience overload for {@link #writeToDataset(SpectralDataset, boolean)}
     * that defaults {@code overwrite} to {@code false}, mirroring
     * Python's keyword-only default.
     *
     * @since 1.1.0
     */
    public void writeToDataset(SpectralDataset dataset) {
        writeToDataset(dataset, false);
    }

    /** Local hex-decoder for the {@code @md5} attribute. Returns
     *  {@code null} if the input is not a 32-char hex string (so the
     *  4-arg constructor falls back to recomputing). */
    private static byte[] parseHexLocal(String hex) {
        if (hex == null || hex.length() != 32) return null;
        byte[] out = new byte[16];
        for (int i = 0; i < 16; i++) {
            int hi = Character.digit(hex.charAt(i * 2), 16);
            int lo = Character.digit(hex.charAt(i * 2 + 1), 16);
            if (hi < 0 || lo < 0) return null;
            out[i] = (byte) ((hi << 4) | lo);
        }
        return out;
    }
}

File: python/src/ttio/genomic/reference_import.py
"""Reference-FASTA value class and embedding helpers.

A ``ReferenceImport`` is the parsed result of a reference-FASTA file
(many short or long chromosome records, no quality scores). It carries
the chromosome names, per-chromosome sequence bytes, and a content-MD5
suitable for the ``@md5`` attribute on
``/study/references/<uri>/`` groups inside a ``.tio`` container.

The same value class is produced by :class:`ttio.importers.fasta.FastaReader`
and consumed by :class:`ttio.exporters.fasta.FastaWriter`, so a
FASTA -> .tio -> FASTA round-trip preserves chromosome names, byte
contents (case-preserving), and MD5.

Cross-language equivalents
--------------------------
Objective-C: ``TTIOReferenceImport`` ·
Java: ``global.thalion.ttio.genomic.ReferenceImport``.

SPDX-License-Identifier: Apache-2.0
"""
from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:  # pragma: no cover
    from pathlib import Path

    from ..io.progress import ProgressSinkLike
    from ..providers.base import StorageGroup
    from ..spectral_dataset import SpectralDataset


__all__ = ["ReferenceImport", "compute_reference_md5"]


def compute_reference_md5(chromosomes: list[str], sequences: list[bytes]) -> bytes:
    """Compute the canonical MD5 over a reference's chromosome set.

    Algorithm (cross-language byte-exact):
        for each (name, seq) in sorted(zip(chromosomes, sequences),
                                       key=name):
            h.update(seq)
        return h.digest()

    Sorting by name makes the MD5 invariant to FASTA record order;
    sequence bytes are concatenated verbatim (case-preserving) without
    any framing. This is the single canonical form used both by the
    REF_DIFF_V2 auto-embed writer (``_reference_md5_for_run``) and by
    the FASTA-import path's ``@md5`` stamping — unified in v1.1.0
    (previously the FASTA-import / public-helper path used a
    name-framed form that disagreed with the writer; see CHANGELOG).

    Returns
    -------
    bytes
        16-byte MD5 digest.
    """
    if len(chromosomes) != len(sequences):
        raise ValueError(
            f"chromosome / sequence length mismatch: "
            f"{len(chromosomes)} names vs {len(sequences)} sequences"
        )
    items = sorted(zip(chromosomes, sequences), key=lambda kv: kv[0])
    h = hashlib.md5()
    for _name, seq in items:
        h.update(seq)
    return h.digest()


@dataclass(slots=True)
class ReferenceImport:
    """Reference-FASTA contents staged for embedding into a ``.tio`` container.

    Parameters
    ----------
    uri : str
        Reference identifier (e.g. ``"GRCh38.p14"``). Becomes the
        sub-group name under ``/study/references/<uri>/``.
    chromosomes : list[str]
        Chromosome names in FASTA file order. The on-disk MD5 is
        order-invariant (see :func:`compute_reference_md5`).
    sequences : list[bytes]
        Per-chromosome sequence bytes, one entry per chromosome,
        case-preserved. Newlines and whitespace stripped.
    md5 : bytes
        16-byte content MD5 (see :func:`compute_reference_md5`). If
        omitted, computed from ``chromosomes`` + ``sequences``.

    Notes
    -----
    A round-trip FASTA -> :class:`FastaReader` -> ``ReferenceImport``
    -> :class:`FastaWriter` -> FASTA preserves byte content
    (including soft-masking via lowercase). The Reference Resolver's
    upper-casing for REF_DIFF_V2 is a separate normalisation that
    happens at decode time.
    """

    uri: str
    chromosomes: list[str]
    sequences: list[bytes]
    md5: bytes = field(default=b"")

    def __post_init__(self) -> None:
        if not self.md5:
            self.md5 = compute_reference_md5(self.chromosomes, self.sequences)
        if len(self.md5) != 16:
            raise ValueError(
                f"md5 must be 16 bytes, got {len(self.md5)}"
            )
        if len(self.chromosomes) != len(self.sequences):
            raise ValueError(
                f"chromosomes / sequences length mismatch: "
                f"{len(self.chromosomes)} vs {len(self.sequences)}"
            )

    @property
    def total_bases(self) -> int:
        """Sum of sequence lengths across all chromosomes."""
        return sum(len(s) for s in self.sequences)

    def chromosome(self, name: str) -> bytes:
        """Return the named chromosome's sequence bytes.

        Raises
        ------
        KeyError
            If ``name`` is not in this reference.
        """
        for n, s in zip(self.chromosomes, self.sequences):
            if n == name:
                return s
        raise KeyError(
            f"chromosome {name!r} not present in reference {self.uri!r} "
            f"(known: {sorted(self.chromosomes)})"
        )

    @classmethod
    def read_from_group(cls, ref_group: "StorageGroup") -> "ReferenceImport":
        """Read an embedded reference from ``/study/references/<uri>/``.

        Inverse of :func:`ttio.spectral_dataset._embed_references_for_runs`.
        Per-chromosome sequences live at
        ``<uri>/chromosomes/<name>/data`` (UINT8); the URI group
        carries ``@reference_uri`` (the canonical URI string) and
        ``@md5`` (32-character lowercase hex). The MD5 is preserved
        verbatim from the on-disk attribute when present, so the
        read-back value carries the same digest bytes the writer used —
        load-bearing for cross-language byte-exact round-trip. When the
        attribute is absent or malformed, the constructor falls back to
        recomputing via :func:`compute_reference_md5`.

        Chromosome names are returned in the order
        :meth:`StorageGroup.child_names` yields them, which for the
        canonical writer is alphabetic (the embed helper sorts before
        persisting).

        Cross-language equivalents
        --------------------------
        Java: ``ReferenceImport.readFromGroup(StorageGroup)``.

        :since: 1.1.0
        """
        from .. import _hdf5_io as io

        # URI: prefer @reference_uri, fall back to the leaf group name.
        uri = io.read_string_attr(ref_group, "reference_uri", default=None)
        if not uri:
            # StorageGroup.name gives the full path for HDF5; take the
            # leaf to mirror Java's refGroup.name() semantics.
            full = ref_group.name
            uri = full.rsplit("/", 1)[-1] if "/" in full else full

        # MD5: read @md5 (lowercase hex string) verbatim. Parse to 16
        # bytes; on absence or malformed input, leave as empty so the
        # ReferenceImport constructor recomputes from sequences.
        md5_bytes = b""
        md5_hex = io.read_string_attr(ref_group, "md5", default=None)
        if md5_hex and len(md5_hex) == 32:
            try:
                md5_bytes = bytes.fromhex(md5_hex)
            except ValueError:
                md5_bytes = b""

        from . import packed_reference

        chrom_names: list[str] = []
        sequences: list[bytes] = []
        chroms_grp = ref_group.open_group("chromosomes")
        try:
            for name in chroms_grp.child_names():
                chrom_grp = chroms_grp.open_group(name)
                try:
                    # Dispatches on layout: data_packed (2-bit + run
                    # mask) when present, legacy raw data otherwise.
                    chrom_names.append(name)
                    sequences.append(
                        packed_reference.read_chromosome_bytes(chrom_grp))
                finally:
                    chrom_grp.close()
        finally:
            chroms_grp.close()

        return cls(
            uri=uri,
            chromosomes=chrom_names,
            sequences=sequences,
            md5=md5_bytes,
        )

    def write_to_dataset(
        self,
        dataset: "SpectralDataset",
        *,
        overwrite: bool = False,
        progress: "ProgressSinkLike | None" = None,
    ) -> None:
        """Embed this reference at ``/study/references/<uri>/``
        inside ``dataset``'s open HDF5 file.

        Layout (cross-language byte-equal — matches Java's
        ``SpectralDataset.embedReferencesForRuns`` and Python's
        :func:`ttio.spectral_dataset._embed_references_for_runs`,
        the canonical writer used by ``embedReference=True`` runs):

        ``/study/references/<uri>/``
          attr ``md5`` : 32-character lowercase hex string.
          attr ``reference_uri`` : the URI string (mirrors the leaf
              path; ``read_from_group`` falls back to the leaf when
              absent, but the canonical writer always emits this).
          ``chromosomes/<name>/`` one sub-group per chromosome, in
              alphabetic order:

            attr ``length`` : int64 sequence length in bytes.
            ``data_packed`` : UINT8 dataset holding the 2-bit + run-mask
                packed stream (``packed_reference`` layout), ZLIB-
                compressed — written when packing beats the raw bytes.
            ``data`` : UINT8 dataset of the chromosome's raw sequence
                bytes (case-preserving, no newlines), ZLIB-compressed —
                the fallback when packing does not win (soft-masked or
                IUPAC-dense sequences), and the only layout pre-change
                readers understand.

        ``@total_bases`` (a v1.1.0-era attribute written here only,
        never by the canonical embed-helper / Java writer) is no
        longer emitted; ``ReferenceImport.read_from_group`` recomputes
        it from the per-chromosome data, so the on-disk attribute
        carried no information that wasn't also derivable.

        Parameters
        ----------
        dataset : SpectralDataset
            Open dataset (writable HDF5 backing).
        overwrite : bool
            If ``True``, replace any existing reference under the same
            URI; if ``False``, raise on collision.
        progress : ProgressSinkLike or None
            Optional runtime progress callback. Fires ``(0, N)`` before
            the embed loop then ``(i+1, N)`` after each contig (N =
            sorted contig count), mirroring Java's
            ``ReferenceImport.writeToDataset(..., ProgressSink)``. A
            bare ``(done, total) -> None`` callable is also accepted.
            Progress is a runtime callback only -- no on-disk change.

        Raises
        ------
        FileExistsError
            If a reference with the same ``uri`` is already embedded
            and ``overwrite`` is ``False``.
        RuntimeError
            If ``dataset``'s storage backend doesn't expose a provider
            (an open dataset always does).
        """
        from .. import _hdf5_io as io
        from ..io.progress import _fire
        from . import packed_reference

        provider = getattr(dataset, "provider", None)
        if provider is None:
            raise RuntimeError(
                "ReferenceImport.write_to_dataset requires a "
                "provider-backed dataset; got "
                f"{type(dataset).__name__} with no .provider."
            )
        root = provider.root_group()
        study = (
            root.open_group("study")
            if root.has_child("study")
            else root.create_group("study")
        )
        references = (
            study.open_group("references")
            if study.has_child("references")
            else study.create_group("references")
        )
        path = f"/study/references/{self.uri}"
        if references.has_child(self.uri):
            if not overwrite:
                raise FileExistsError(
                    f"reference {self.uri!r} already embedded at {path}; "
                    f"pass overwrite=True to replace."
                )
            references.delete_child(self.uri)

        # Write through the StorageGroup protocol so the attribute layout
        # (NULLTERM fixed-length strings via write_fixed_string_attr,
        # int64 via write_int_attr) matches the canonical embed-helper
        # writer and Java's embedReferencesForRuns byte-for-byte.
        ref_grp = references.create_group(self.uri)
        io.write_fixed_string_attr(ref_grp, "md5", self.md5.hex())
        io.write_fixed_string_attr(ref_grp, "reference_uri", self.uri)
        chroms_grp = ref_grp.create_group("chromosomes")
        # Sort alphabetically so the on-disk child order matches what
        # the canonical writer emits (read_from_group surfaces names in
        # on-disk order).
        sorted_names = sorted(self.chromosomes)
        total = len(sorted_names)
        # Mirror Java ReferenceImport.writeToDataset: (0, N) then
        # (i+1, N) per contig. Progress is a runtime callback only.
        _fire(progress, 0, total)
        for i, name in enumerate(sorted_names):
            seq = self.chromosome(name)
            c = chroms_grp.create_group(name)
            io.write_int_attr(c, "length", len(seq))
            packed_reference.write_chromosome_dataset(c, seq)
            _fire(progress, i + 1, total)
